### PR TITLE
Improving LM training (custom optimizer, custom scheduler, Transformer LM, etc)

### PR DESCRIPTION
**Pull request by ShigekiKarita**
- [x] implement cached encoder for Transformer LM (benchmark)
- [x] implement pos enc-free Transformer LM https://arxiv.org/pdf/1905.04226.pdf
- [x] espnet.scheduler (model/backend-generic impl, noam, cosine warmup/annealing)
- [x] espnet.optimizer (model/backend-generic impl, sgd, adam, adadelta)
- [x] --accum-grad
- [x] chainer support (except for Transformer LM)
- [x] update README/RESULTS
- [x] uploaded model and recov_wav.sh

Notes:
  - Irie config (no pos enc) https://github.com/rwth-i6/returnn-experiments/blob/master/2019-lm-transformers/librispeech/bpe_10k/transfo_24_d00.2048_512.sgd.lr1.8_heads.config
  - GPT2 config (cosine annealing) https://github.com/NVIDIA/Megatron-LM#gpt2-pretraining

See also #1240

### CODE REVIEW

**Comment by ShigekiKarita** [comment]
benchmark_encoder

**Comment by ShigekiKarita** [comment]
I found that running TransformerLM exp on LibriSpeech will take **3 months**... So I would like to finish this PR before updating RESULTS.

File: espnet/scheduler/chainer.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+"""Chainer optimizer schdulers."""
```
[the remark this answers is not in the document]
Comment (ShigekiKarita) [reply espnet/scheduler/chainer.py]: I wanna make these schedulers and cmd parsers **universal** for all the backends. I consider that unifying everything but separating schedulers in each backends. Maybe schedulers in chainer and pytorch can do more than my impl but the current ESPnet training does not require it.

File: egs/librispeech/asr1/run.sh [hunk 2]
```diff
@@ -11,6 +11,7 @@ backend=pytorch
 stage=-1       # start from -1 if you need to start from data download
 stop_stage=100
 ngpu=4         # number of gpus ("0" uses cpu, otherwise use gpu)
+nj=32
```
[the remark this answers is not in the document]
Comment (ShigekiKarita) [reply egs/librispeech/asr1/run.sh]: I agree. When I test the recipe in my small PC, the configurable `nj` really helps :sob:. However, the other recipes might not have such a high load average problem on this (because LibriSpeech is the biggest dataset in ESPnet.)

**Comment by kan-bayashi** [comment]
Is it time to review?

**Comment by ShigekiKarita** [comment]
Yes!

2019年10月12日(土) 1:16 Tomoki Hayashi <notifications@github.com>:

File: egs/librispeech/asr1/conf/tuning/lm_transformer.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# see https://arxiv.org/pdf/1905.04226.pdf
```
Comment (sw005320) [inline egs/librispeech/asr1/conf/tuning/lm_transformer.yaml]: Can you make it more structured by separating the topology definition, optimization related etc, similar to the e2e ASR config?

Comment (kan-bayashi) [reply egs/librispeech/asr1/conf/tuning/lm_transformer.yaml]: I like adding comments in header.
(Intension of configuration, requirements, how long does it take....)
If you add brief comment, that is nice.

File: egs/librispeech/asr1/run.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (kan-bayashi) [reply egs/librispeech/asr1/run.sh]: in tts recipes, `nj` is hyperparameter :)

File: espnet/bin/lm_train.py [hunk 4]
```diff
@@ -62,28 +65,31 @@ def get_parser():
...
+    parser.add_argument('--accum-grad', type=int, default=1,
+                        help='Number of gradient accumueration')
     parser.add_argument('--epoch', '-e', type=int, default=20,
                         help='Number of sweeps over the dataset to train')
     parser.add_argument('--early-stop-criterion', default='validation/main/loss', type=str, nargs='?',
                         help="Value to monitor to trigger an early stopping of the training")
     parser.add_argument('--patience', default=3, type=int, nargs='?',
                         help="Number of epochs to wait without improvement before stopping the training")
+    parser.add_argument('--scalers', default=None, action="append", type=lambda kv: kv.split("="),
+                        help='optimizer schedulers, e.g., "--scalers lr=noam --lr-noam-warmup 1000".')
```
Comment (kan-bayashi) [inline espnet/bin/lm_train.py]: Why the name is `scalers` but not `schedulers`?
Could you teach me what stand for `lr=`?

File: espnet/optimizer/adaptor.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
+"""Import optimizer class dynamically."""
+import argparse
+
+from espnet.optimizer.parser import OPTIMIZER_PARSER_DICT
+from espnet.utils.dynamic_import import dynamic_import
+from espnet.utils.fill_missing_args import fill_missing_args
+
+
+class OptimizerAdaptorInterface:
```
Comment (kan-bayashi) [inline espnet/optimizer/adaptor.py]: According to #943, we decided to use `OptimizerFactoryInterface`.
Why did you change your mind?

File: espnet/optimizer/adaptor.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            new Optimizer
+
+        """
+        return self.builder(target, args)
+
+    def build(self, target, **kwargs):
+        """Initialize optimizer with python-level args.
+
+        Args:
+            parameters: for pytorch `model.parameters()`,
```
Comment (kan-bayashi) [inline espnet/optimizer/adaptor.py]:
```suggestion
            target: for pytorch `model.parameters()`,
```

File: espnet/optimizer/adaptor.py (hunk 5 above)
Comment (kan-bayashi) [reply espnet/optimizer/adaptor.py]: And I'm still not clear what this class is used for.
Could you tell me how to use `OptimizerAdaptorInterface` and `FunctionalOptimizerAdaptorInterface` properly?

File: espnet/scheduler/scaler.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+    def get_arg(self, name):
+        """Get argument without prefix."""
+        return getattr(self.args, f"{self.key}_{self.alias}_{name}")
+
+    @classmethod
+    def add_arguments(cls, key: str, parser: argparse.ArgumentParser):
+        """Add arguments for CLI."""
+        group = parser.add_argument_group(f"{cls.alias} scaler")
+        cls._add_arguments(_PrefixParser(parser=group, prefix=f"--{key}-{cls.alias}-"))
```
Comment (kan-bayashi) [inline espnet/scheduler/scaler.py]: Could you tell me why you nest function?

Comment (kan-bayashi) [reply espnet/scheduler/scaler.py]: Now I got it.
But I'm not sure it is really needed.

**Comment by ShigekiKarita** [comment]
> This is very cool.
> Thanks, @ShigekiKarita.
> Did you observe some improvement by using the transformer LM from LSTMLM?

No yet.... I am running 6 and 12 layers model for testing but Irie-san did 96 layers to beat RNNLM. At this time, on LibriSpeech:

- our best RNNLM (similar to Irie-san config): train ppl 32.4, valid ppl 35.6
- my best 6 layers TransformerLM (noam, lr=1e-4, bs=128, 8 gpu, at epoch25): train ppl 40.5, valid ppl 42.0
- my best 12 layers TransformerLM (noam, lr=1e-4, bs=32, 8 gpu, at epoch 4): train ppl 38.8, valid ppl 40.1

As you know in Transformer ASR, they are still improving... Let's see the result in next week...

File: egs/librispeech/asr1/conf/tuning/lm_transformer.yaml (hunk 3 above)
Comment (ShigekiKarita) [reply egs/librispeech/asr1/conf/tuning/lm_transformer.yaml]: I see. I will make sections with spaces and add more comments for choices and requirements.

File: espnet/bin/lm_train.py (hunk 4 above)
Comment (ShigekiKarita) [reply espnet/bin/lm_train.py]: Yeah, the name `schedulers` is better. Actually, you can specify any hyperparameters in optimizer like `--schedulers momentum=noam` then its hyper-hyperparameter is `--momentum-warmup 5000`

Comment (ShigekiKarita) [reply espnet/bin/lm_train.py]: I will add this behaviour to specify multiple scheduling in the help.

File: espnet/optimizer/adaptor.py (hunk 5 above)
Comment (ShigekiKarita) [reply espnet/optimizer/adaptor.py]: Oh, I almost forget it. I will use `OptimizerFactoryInterface` instead of  `OptimizerAdaptorInterface`. First, I planned to use `OptimizerXXXInterface` as usual (or seen in #943) but I jjust tried the functional way is much easier to define the parsers and chainer/pytorch adaptors for checking which one is better. I have no answer yet. Do you think which one is better?

File: espnet/scheduler/scaler.py (hunk 7 above)
Comment (ShigekiKarita) [reply espnet/scheduler/scaler.py]: This is related to your previous question. I just want to support multiple schedules for multiple hyperparameters (as SpecAugument paper did). Therefore, I need to rename args like `--warmup` for `lr` and `weight_decay` to `--lr-warmup` and `--weight_decay-warmup`

File: espnet/optimizer/adaptor.py (hunk 5 above)
Comment (ShigekiKarita) [reply espnet/optimizer/adaptor.py]: Another reason why I used a functional way is that #943 only uses `@classmethod` to define `OptimizerFactoryInterface`. That is essentially the same as a pure function because it does not require any states in `self`. And for Python syntax, the function is much simpler than class.

**Comment by ShigekiKarita** [comment]
Sorry for being absent. I could not access my server for a while. The Transformer LM with cosine annealing seems to get lower PPL than the baseline RNNLM now. (I updated the table above)

**Comment by ShigekiKarita** [comment]
WIP WER by beam=10, ASR model at google drive

| LM | dev clean | dev other | test clean | test other |
|:----:|--:|--:|--:|--:|
|Transformer (cosine ep8 wip) | 2.3 | 6.0 | 2.6 | 6.1 |
|LSTM (irielm at google drive, model averaged) | 2.3 | 5.9 | 2.7 | 6.0 |

I found that the decoding speed of Transformer LM and LSTM LM is almost the same there.

**Comment by sw005320** [comment]
maybe it's time to review it?

**Comment by ShigekiKarita** [comment]
Sorry for long time absent. I would like to finish this PR today. First of all, I would like to share my final result for someone who wants this best LM model ever in ESPnet (NOTE that the ASR model is the same before).

(TODO: update RESULTS.md and README.md)

  - Model files (archived to model.tar.gz by `$ pack_model.sh`)
    - model link: https://drive.google.com/open?id=17cOOSHHMKI82e1MXj4r2ig8gpGCRmG2p
    - training config file: `./conf/train.yaml`
    - decoding config file: `./conf/decode.yaml`
    - cmvn file: `./data/train_960/cmvn.ark`
    - e2e file: `./librispeech.transformer.v1/exp/train_960_pytorch_train_pytorch_transformer.v1_aheads8_batch-bins15000000_specaug/results/model.val5.avg.best`
    - e2e JSON file: `./librispeech.transformer.v1/exp/train_960_pytorch_train_pytorch_transformer.v1_aheads8_batch-bins15000000_specaug/results/model.json`
    - lm file: `./exp/train_rnnlm_pytorch_lm_transformer_cosine_batchsize32_lr1e-4_layer16_unigram5000_ngpu4/rnnlm.model.best`
    - lm JSON file: `./exp/train_rnnlm_pytorch_lm_transformer_cosine_batchsize32_lr1e-4_layer16_unigram5000_ngpu4/model.json`
  - Results (paste them by yourself or obtained by `$ pack_model.sh --results <results>`)
```
./exp/train_rnnlm_pytorch_lm_transformer_cosine_batchsize32_lr1e-4_layer16_unigram5000_ngpu4/decode_dev_clean_decode_ep43/result.wrd.txt
|    SPKR            |    # Snt        # Wrd     |    Corr            Sub           Del           Ins           Err         S.Err     |
|    Sum/Avg         |    2703         54402     |    98.1            1.7           0.2           0.2           2.1          26.9     |
./exp/train_rnnlm_pytorch_lm_transformer_cosine_batchsize32_lr1e-4_layer16_unigram5000_ngpu4/decode_dev_other_decode_ep43/result.wrd.txt
|    SPKR            |    # Snt        # Wrd     |    Corr            Sub           Del           Ins           Err         S.Err     |
|    Sum/Avg         |    2864         50948     |    95.3            4.2           0.5           0.6           5.3          43.8     |
./exp/train_rnnlm_pytorch_lm_transformer_cosine_batchsize32_lr1e-4_layer16_unigram5000_ngpu4/decode_test_clean_decode_ep43/result.wrd.txt
|    SPKR            |    # Snt         # Wrd     |    Corr           Sub           Del            Ins           Err         S.Err     |
|    Sum/Avg         |    2620          52576     |    97.8           1.9           0.2            0.3           2.5          28.3     |
./exp/train_rnnlm_pytorch_lm_transformer_cosine_batchsize32_lr1e-4_layer16_unigram5000_ngpu4/decode_test_other_decode_ep43/result.wrd.txt
|    SPKR            |    # Snt         # Wrd     |    Corr           Sub           Del            Ins           Err         S.Err     |
|    Sum/Avg         |    2939          52343     |    95.1           4.3           0.6            0.6           5.5          46.7     |
```

**Comment by ShigekiKarita** [comment]
And I'm still decoding with weight averaged models. If they are better, I will update the results and models.

File: espnet/optimizer/adaptor.py (hunk 5 above)
Comment (ShigekiKarita) [reply espnet/optimizer/adaptor.py]: @kan-bayashi I finally rewrite everything from functional style to OOP style

**Comment by ShigekiKarita** [comment]
I have finished refactoring for review comments. Let me wait CI success and averaged model WERs.

**Comment by ShigekiKarita** [comment]
Done. I believe this is ready for merge.

PS: I found model averaging on Transformer LM does not reduce WERs. So I modified run.sh not to average LMs by default.

**Comment by ShigekiKarita** [comment]
I also found that the final Transformer LM is +300MB larger but faster than LSTM LM :face_with_thermometer:  
- 12.880 sec, `time ../../../utils/recog_wav.sh --stage 3 --decode-config ./conf/decode_beam-size4.yaml  --models librispeech.transformer.v1.transformerlm.v1 ~/Desktop/wavs/1089-134686-0000.wav`
- 14.125 sec, `time ../../../utils/recog_wav.sh --stage 3 --decode-config ./conf/decode_beam-size4.yaml  --models librispeech.transformer.v1 ~/Desktop/wavs/1089-134686-0000.wav`

**Comment by sw005320** [comment]
@kan-bayashi, can you take a look at the modification?

**Comment by ShigekiKarita** [comment]
The main modification is making OptimizerFactory (espnet.optimizer.{factory,chainer,pytorch,parser}) into OOP style from functional style. See this commit https://github.com/espnet/espnet/pull/1246/commits/cabc1a51ff5476b387dd8012a92ba315fca0561b

File: egs/librispeech/asr1/run.sh [hunk 8]
```diff
@@ -287,7 +294,8 @@ if [ ${stage} -le 5 ] && [ ${stop_stage} -ge 5 ]; then
             --recog-json ${feat_recog_dir}/split${nj}utt/data_${bpemode}${nbpe}.JOB.json \
             --result-label ${expdir}/${decode_dir}/data.JOB.json \
             --model ${expdir}/results/${recog_model}  \
-            --rnnlm ${lmexpdir}/${lang_model}
+            --rnnlm ${lmexpdir}/${lang_model} \
+	    --api v2
```
Comment (sw005320) [inline egs/librispeech/asr1/run.sh]: please adjust the indent.

Comment (ShigekiKarita) [reply egs/librispeech/asr1/run.sh]: OK

File: egs/librispeech/asr1/conf/tuning/lm_transformer.yaml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+pos-enc: none
+unit: 2048
+
+# minibatch related
+batchsize: 32
+maxlen: 40
+
+# optimization related
+opt: adam
+scalers: lr=cosine
```
Comment (kan-bayashi) [inline egs/librispeech/asr1/conf/tuning/lm_transformer.yaml]: schedulers?

File: egs/mini_an4/asr1/conf/lm_transformer.yaml [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+# optimization related
+opt: adam
+schedulers: lr=cosine
+dropout-rate: 0.0
+epoch: 3
+gradclip: 1.0
+lr: 1e-4
+lr-cosine-total: 3
+lr-cosine-warmup: 2
```
Comment (kan-bayashi) [inline egs/mini_an4/asr1/conf/lm_transformer.yaml]: Let me confirm the usage.
If we specify lr2=cosine, we can specify lr2-cosine-total. Is my understanding correct?

File: egs/librispeech/asr1/RESULTS.md [hunk 11]
```diff
@@ -1,4 +1,42 @@
...
+- Environments
+  - date: `Tue Feb  4 14:50:50 JST 2020`
+  - python version: `3.7.3 (default, Mar 27 2019, 22:11:17)  [GCC 7.3.0]`
+  - espnet version: `espnet 0.6.0`
+  - chainer version: `chainer 6.0.0`
+  - pytorch version: `pytorch 1.0.1.post2`
+  - Git hash: `83799e69a0269450587a6857882c73bfb27551d5`
+  - Commit date: `Tue Feb 4 14:21:11 2020 +0900`
+
+- Model files (archived to model.tar.gz by `$ pack_model.sh`)
```
[the remark this answers is not in the document]
Comment (ShigekiKarita) [reply egs/librispeech/asr1/RESULTS.md]: OK

File: egs/librispeech/asr1/conf/tuning/lm_transformer.yaml (hunk 9 above)
Comment (ShigekiKarita) [reply egs/librispeech/asr1/conf/tuning/lm_transformer.yaml]: OK

**Comment by kan-bayashi** [comment]
@ShigekiKarita We should include sentence piece models. So you can specify `data/lang_char` directory for `--dict`.

File: egs/mini_an4/asr1/conf/lm_transformer.yaml (hunk 10 above)
Comment (ShigekiKarita) [reply egs/mini_an4/asr1/conf/lm_transformer.yaml]: yes only if `lr2` is available as optimizer's hyperparam and `total` in the scheduler. so it should be `<optimizer param>-<scheduler-name>-<scheduler-param>`.

**Comment by ShigekiKarita** [comment]
Thanks for review comments. Done!

File: egs/mini_an4/asr1/conf/lm_transformer.yaml (hunk 10 above)
Comment (ShigekiKarita) [reply egs/mini_an4/asr1/conf/lm_transformer.yaml]: I will add this to lm_train.py

**Comment by ShigekiKarita** [comment]
I believe done!

2020年2月5日(水) 13:06 Tomoki Hayashi <notifications@github.com>:

**Comment by kan-bayashi** [comment]
Thanks a lot!
